MySQL server can't start

I've found a bug on building the Dockerfile from [latest Dockerfile](https://github.com/ZoneMinder/zmdockerfiles/blob/master/development/ubuntu/xenial/Dockerfile). MySQL server can't start. See /var/log/mysql/error.log:
```
2018-09-29T09:39:33.225679Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
2018-09-29T09:39:34.532144Z 0 [Warning] InnoDB: New log files created, LSN=45790
2018-09-29T09:39:34.729063Z 0 [Warning] InnoDB: Creating foreign key constraint system tables.
2018-09-29T09:39:34.802924Z 0 [Warning] No existing UUID has been found, so we assume that this is the first time that this server has been started. Generating a new UUID: 93125299-c3cb-11e8-a4bd-0242ac110003.
2018-09-29T09:39:34.804480Z 0 [Warning] Gtid table is not ready to be used. Table 'mysql.gtid_executed' cannot be opened.
2018-09-29T09:39:34.805052Z 1 [Warning] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option.
2018-09-29T09:48:53.259398Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
2018-09-29T09:48:53.259449Z 0 [Warning] 'NO_AUTO_CREATE_USER' sql mode was not set.
2018-09-29T09:48:53.260379Z 0 [Note] /usr/sbin/mysqld (mysqld 5.7.23-0ubuntu0.16.04.1) starting as process 411 ...
2018-09-29T09:48:53.264233Z 0 [Note] InnoDB: PUNCH HOLE support available
2018-09-29T09:48:53.264266Z 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
2018-09-29T09:48:53.264275Z 0 [Note] InnoDB: Uses event mutexes
2018-09-29T09:48:53.264283Z 0 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier
2018-09-29T09:48:53.264290Z 0 [Note] InnoDB: Compressed tables use zlib 1.2.8
2018-09-29T09:48:53.264298Z 0 [Note] InnoDB: Using Linux native AIO
2018-09-29T09:48:53.264445Z 0 [Note] InnoDB: Number of pools: 1
2018-09-29T09:48:53.264515Z 0 [Note] InnoDB: Using CPU crc32 instructions
2018-09-29T09:48:53.265472Z 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M
2018-09-29T09:48:53.269914Z 0 [Note] InnoDB: Completed initialization of buffer pool
2018-09-29T09:48:53.271294Z 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
2018-09-29T09:48:53.281341Z 0 [Note] InnoDB: The first innodb_system data file 'ibdata1' did not exist. A new tablespace will be created!
2018-09-29T09:48:53.281498Z 0 [Note] InnoDB: Setting file './ibdata1' size to 12 MB. Physically writing the file full; Please wait ...
2018-09-29T09:48:53.795953Z 0 [Note] InnoDB: File './ibdata1' size is now 12 MB.
2018-09-29T09:48:53.796210Z 0 [Note] InnoDB: Setting log file ./ib_logfile101 size to 48 MB
2018-09-29T09:48:54.240885Z 0 [Note] InnoDB: Setting log file ./ib_logfile1 size to 48 MB
2018-09-29T09:48:55.120255Z 0 [Note] InnoDB: Renaming log file ./ib_logfile101 to ./ib_logfile0
2018-09-29T09:48:55.120363Z 0 [Warning] InnoDB: New log files created, LSN=45790
2018-09-29T09:48:55.120392Z 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2018-09-29T09:48:55.120439Z 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2018-09-29T09:48:55.276016Z 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
2018-09-29T09:48:55.276193Z 0 [Note] InnoDB: Doublewrite buffer not found: creating new
2018-09-29T09:48:55.576817Z 0 [Note] InnoDB: Doublewrite buffer created
2018-09-29T09:48:55.580019Z 0 [Note] InnoDB: 96 redo rollback segment(s) found. 96 redo rollback segment(s) are active.
2018-09-29T09:48:55.580045Z 0 [Note] InnoDB: 32 non-redo rollback segment(s) are active.
2018-09-29T09:48:55.580140Z 0 [Warning] InnoDB: Creating foreign key constraint system tables.
2018-09-29T09:48:55.688122Z 0 [Note] InnoDB: Foreign key constraint system tables created
2018-09-29T09:48:55.688206Z 0 [Note] InnoDB: Creating tablespace and datafile system tables.
2018-09-29T09:48:55.788449Z 0 [Note] InnoDB: Tablespace and datafile system tables created.
2018-09-29T09:48:55.788581Z 0 [Note] InnoDB: Creating sys_virtual system tables.
2018-09-29T09:48:55.877285Z 0 [Note] InnoDB: sys_virtual table created
2018-09-29T09:48:55.877433Z 0 [Note] InnoDB: Waiting for purge to start
2018-09-29T09:48:55.927657Z 0 [Note] InnoDB: 5.7.23 started; log sequence number 0
2018-09-29T09:48:55.928042Z 0 [Note] Plugin 'FEDERATED' is disabled.
mysqld: Table 'mysql.plugin' doesn't exist
2018-09-29T09:48:55.928243Z 0 [ERROR] Can't open the mysql.plugin table. Please run mysql_upgrade to create it.
2018-09-29T09:48:55.928581Z 0 [Note] Salting uuid generator variables, current_pid: 411, server_start_time: 1538214533, bytes_sent: 0, 
2018-09-29T09:48:55.928622Z 0 [Note] Generated uuid: 'e1874fc3-c3cc-11e8-b86e-0242ac130002', server_start_time: 115686216966294560, bytes_sent: 44727424
2018-09-29T09:48:55.928635Z 0 [Warning] No existing UUID has been found, so we assume that this is the first time that this server has been started. Generating a new UUID: e1874fc3-c3cc-11e8-b86e-0242ac130002.
2018-09-29T09:48:56.044738Z 0 [Warning] Gtid table is not ready to be used. Table 'mysql.gtid_executed' cannot be opened.
2018-09-29T09:48:56.044812Z 0 [Warning] Failed to set up SSL because of the following SSL library error: SSL context is not usable without certificate and private key
2018-09-29T09:48:56.044827Z 0 [Note] Server hostname (bind-address): '127.0.0.1'; port: 3306
2018-09-29T09:48:56.044837Z 0 [Note]   - '127.0.0.1' resolves to '127.0.0.1';
2018-09-29T09:48:56.044863Z 0 [Note] Server socket created on IP: '127.0.0.1'.
2018-09-29T09:48:56.045899Z 0 [Warning] Failed to open optimizer cost constant tables

2018-09-29T09:48:56.046013Z 0 [ERROR] Fatal error: Can't open and lock privilege tables: Table 'mysql.user' doesn't exist
2018-09-29T09:48:56.046035Z 0 [ERROR] Fatal error: Failed to initialize ACL/grant/time zones structures or failed to remove temporary table files.
2018-09-29T09:48:56.046085Z 0 [ERROR] Aborting

2018-09-29T09:48:56.046116Z 0 [Note] Binlog end
2018-09-29T09:48:56.046160Z 0 [Note] Shutting down plugin 'ngram'
2018-09-29T09:48:56.046176Z 0 [Note] Shutting down plugin 'partition'
2018-09-29T09:48:56.046185Z 0 [Note] Shutting down plugin 'BLACKHOLE'
2018-09-29T09:48:56.046195Z 0 [Note] Shutting down plugin 'ARCHIVE'
2018-09-29T09:48:56.046204Z 0 [Note] Shutting down plugin 'MEMORY'
2018-09-29T09:48:56.046213Z 0 [Note] Shutting down plugin 'INNODB_SYS_VIRTUAL'
2018-09-29T09:48:56.046223Z 0 [Note] Shutting down plugin 'INNODB_SYS_DATAFILES'
2018-09-29T09:48:56.046232Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESPACES'
2018-09-29T09:48:56.046241Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN_COLS'
2018-09-29T09:48:56.046251Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN'
2018-09-29T09:48:56.046259Z 0 [Note] Shutting down plugin 'INNODB_SYS_FIELDS'
2018-09-29T09:48:56.046269Z 0 [Note] Shutting down plugin 'INNODB_SYS_COLUMNS'
2018-09-29T09:48:56.046278Z 0 [Note] Shutting down plugin 'INNODB_SYS_INDEXES'
2018-09-29T09:48:56.046287Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESTATS'
2018-09-29T09:48:56.046296Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLES'
2018-09-29T09:48:56.046306Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_TABLE'
2018-09-29T09:48:56.046315Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_CACHE'
2018-09-29T09:48:56.046324Z 0 [Note] Shutting down plugin 'INNODB_FT_CONFIG'
2018-09-29T09:48:56.046332Z 0 [Note] Shutting down plugin 'INNODB_FT_BEING_DELETED'
2018-09-29T09:48:56.046341Z 0 [Note] Shutting down plugin 'INNODB_FT_DELETED'
2018-09-29T09:48:56.046350Z 0 [Note] Shutting down plugin 'INNODB_FT_DEFAULT_STOPWORD'
2018-09-29T09:48:56.046359Z 0 [Note] Shutting down plugin 'INNODB_METRICS'
2018-09-29T09:48:56.046368Z 0 [Note] Shutting down plugin 'INNODB_TEMP_TABLE_INFO'
2018-09-29T09:48:56.046377Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_POOL_STATS'
2018-09-29T09:48:56.046386Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE_LRU'
2018-09-29T09:48:56.046394Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE'
2018-09-29T09:48:56.046403Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX_RESET'
2018-09-29T09:48:56.046412Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX'
2018-09-29T09:48:56.046424Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM_RESET'
2018-09-29T09:48:56.046434Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM'
2018-09-29T09:48:56.046443Z 0 [Note] Shutting down plugin 'INNODB_CMP_RESET'
2018-09-29T09:48:56.046452Z 0 [Note] Shutting down plugin 'INNODB_CMP'
2018-09-29T09:48:56.046461Z 0 [Note] Shutting down plugin 'INNODB_LOCK_WAITS'
2018-09-29T09:48:56.046470Z 0 [Note] Shutting down plugin 'INNODB_LOCKS'
2018-09-29T09:48:56.046478Z 0 [Note] Shutting down plugin 'INNODB_TRX'
2018-09-29T09:48:56.046491Z 0 [Note] Shutting down plugin 'InnoDB'
2018-09-29T09:48:56.046544Z 0 [Note] InnoDB: FTS optimize thread exiting.
2018-09-29T09:48:56.046585Z 0 [Note] InnoDB: Starting shutdown...
2018-09-29T09:48:56.146769Z 0 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
2018-09-29T09:48:56.146858Z 0 [Note] InnoDB: Buffer pool(s) dump completed at 180929 12:48:56
2018-09-29T09:48:57.978041Z 0 [Note] InnoDB: Shutdown completed; log sequence number 1209961
2018-09-29T09:48:57.980898Z 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
2018-09-29T09:48:57.980947Z 0 [Note] Shutting down plugin 'MRG_MYISAM'
2018-09-29T09:48:57.980967Z 0 [Note] Shutting down plugin 'MyISAM'
2018-09-29T09:48:57.980993Z 0 [Note] Shutting down plugin 'CSV'
2018-09-29T09:48:57.981010Z 0 [Note] Shutting down plugin 'PERFORMANCE_SCHEMA'
2018-09-29T09:48:57.981072Z 0 [Note] Shutting down plugin 'sha256_password'
2018-09-29T09:48:57.981088Z 0 [Note] Shutting down plugin 'mysql_native_password'
2018-09-29T09:48:57.981342Z 0 [Note] Shutting down plugin 'binlog'
2018-09-29T09:48:57.982144Z 0 [Note] /usr/sbin/mysqld: Shutdown complete

```

I'm still looking into this, but I don't think the problem is caused by a missing `mysql_install_db`.
True, its deprecated and we need to update it, but I'm not sure that is the root cause of the issue you reported.

I've fired up a vm matching your distro, and that command does exist:
```
abauer@vXenial:~$ mysqld --version
mysqld  Ver 5.7.23-0ubuntu0.16.04.1 for Linux on x86_64 ((Ubuntu))
abauer@vXenial:~$ type mysql_install_db
mysql_install_db is /usr/bin/mysql_install_db
```

From your log output, this line seems to be relevant:
```
2018-09-29T09:48:55.928243Z 0 [ERROR] Can't open the mysql.plugin table. Please run mysql_upgrade to create it.
```
Still digging....

--- a/utils/entrypoint.sh
+++ b/utils/entrypoint.sh
@@ -212,7 +212,7 @@ start_mysql () {
 
     if [ "$(mysql_datadir_exists)" -eq "0" ]; then
         echo " * First run of MYSQL, initializing DB."
-        if [ "$(use_mysql_install_db)" -eq "0" ]; then
+        if [ "$(use_mysql_install_db)" -eq "1" ]; then
             ${MYSQL_INSTALL_DB} --user=mysql --datadir=/var/lib/mysql/ > /dev/null 2>&1
         else
             ${MYSQLD} --initialize-insecure --user=mysql --datadir=/var/lib/mysql/ > /dev/null 2>&1

--- a/utils/entrypoint.sh
+++ b/utils/entrypoint.sh
@@ -74,7 +74,7 @@ initialize () {
         fi
     done
 
-    for FILE in $ZMCONF $ZMPKG $ZMCREATE $PHPINI $HTTPBIN $MYSQLD; do 
+    for FILE in $ZMCONF $ZMPKG $ZMCREATE $PHPINI $HTTPBIN $MYSQLD; do
         if [ -z $FILE ]; then
             echo
             echo "FATAL: This script was unable to determine one or more critical files. Cannot continue."
@@ -183,8 +183,8 @@ use_mysql_install_db () {
     MYSQL_INSTALL_DB=$(type mysql_install_db)
     local result="$?"
 
-    if [ "$result" -eq "0" ] && [ -n $MYSQL_INSTALL_DB  ]; then
-        local contents=$(file -b $MYSQL_INSTALL_DB)
+    if [ "$result" -eq "0" ] && [ -n "$MYSQL_INSTALL_DB"  ]; then
+        local contents=$(file -b "$MYSQL_INSTALL_DB")
         if [[ "$contents" =~ "*ASCII text executable*" ]]; then
             echo "1" # mysql_install_db is a shell script
         else
@@ -237,10 +237,10 @@ start_mysql () {
         echo " * MySQL database server already running."
     fi
 
-    mysqlpid=`cat "$mypidfile" 2>/dev/null`    
+    mysqlpid=`cat "$mypidfile" 2>/dev/null`
 }
 
-# Check the status of the remote mysql server using /etc/php/7.2/apache2/php.inisupplied credentials
+# Check the status of the remote mysql server using supplied credentials
 chk_remote_mysql () {
     if [ $remoteDB -eq "1" ]; then
         echo -n " * Looking for remote database server"

--- a/utils/entrypoint.sh
+++ b/utils/entrypoint.sh
@@ -14,7 +14,7 @@
 initialize () {
 
     # Check to see if this script has access to all the commands it needs
-    for CMD in cat grep install ln my_print_defaults mysql mysqladmin mysqld_safe mysql_install_db mysqlshow sed sleep su tail usermod; do
+    for CMD in cat grep install ln my_print_defaults mysql mysqladmin mysqld_safe mysqlshow sed sleep su tail usermod head file; do
       type $CMD &> /dev/null
 
       if [ $? -ne 0 ]; then
@@ -23,6 +23,14 @@ initialize () {
         echo
         exit 1
       fi
+    done
+
+    # Look in common places for the mysqld/MariaDB executable
+    for FILE in "/usr/sbin/mysqld" "/usr/libexec/mysqld" "/usr/local/sbin/mysqld" "/usr/local/libexec/mysqld"; do
+        if [ -f $FILE ]; then
+            MYSQLD=$FILE
+            break
+        fi
     done
 
     # Look in common places for the apache executable commonly called httpd or apache2
@@ -66,7 +74,7 @@ initialize () {
         fi
     done
 
-    for FILE in $ZMCONF $ZMPKG $ZMCREATE $PHPINI $HTTPBIN; do 
+    for FILE in $ZMCONF $ZMPKG $ZMCREATE $PHPINI $HTTPBIN $MYSQLD; do 
         if [ -z $FILE ]; then
             echo
             echo "FATAL: This script was unable to determine one or more critical files. Cannot continue."
@@ -79,6 +87,7 @@ initialize () {
             echo "Path to zm_create.sql: ${ZMCREATE}"
             echo "Path to php.ini: ${PHPINI}"
             echo "Path to Apache executable: ${HTTPBIN}"
+            echo "Path to Mysql executable: ${MYSQLD}"
             echo
             exit 98
         fi
@@ -167,6 +176,25 @@ zm_db_exists() {
     fi
 }
 
+# The secret sauce to determine wether to use mysql_install_db
+# or mysqld --initialize seems to be wether mysql_install_db is a shell
+# script or a binary executable
+use_mysql_install_db () {
+    MYSQL_INSTALL_DB=$(type mysql_install_db)
+    local result="$?"
+
+    if [ "$result" -eq "0" ] && [ -n $MYSQL_INSTALL_DB  ]; then
+        local contents=$(file -b $MYSQL_INSTALL_DB)
+        if [[ "$contents" =~ "*ASCII text executable*" ]]; then
+            echo "1" # mysql_install_db is a shell script
+        else
+            echo "0" # mysql_install_db is a binary
+        fi
+    else
+        echo "0" # mysql_install_db does not exist
+    fi
+}
+
 # mysql service management
 start_mysql () {
     # determine if we are running mariadb or mysql then guess pid location
@@ -184,7 +212,11 @@ start_mysql () {
 
     if [ "$(mysql_datadir_exists)" -eq "0" ]; then
         echo " * First run of MYSQL, initializing DB."
-        mysql_install_db --user=mysql --datadir=/var/lib/mysql/ > /dev/null 2>&1
+        if [ "$(use_mysql_install_db)" -eq "0" ]; then
+            ${MYSQL_INSTALL_DB} --user=mysql --datadir=/var/lib/mysql/ > /dev/null 2>&1
+        else
+            ${MYSQLD} --initialize-insecure --user=mysql --datadir=/var/lib/mysql/ > /dev/null 2>&1
+        fi
     elif [ -e ${mypidsocklock} ]; then
         echo " * Removing stale lock file"
         rm -f ${mypidsocklock}
@@ -208,7 +240,7 @@ start_mysql () {
     mysqlpid=`cat "$mypidfile" 2>/dev/null`    
 }
 
-# Check the status of the remote mysql server using supplied credentials
+# Check the status of the remote mysql server using /etc/php/7.2/apache2/php.inisupplied credentials
 chk_remote_mysql () {
     if [ $remoteDB -eq "1" ]; then
         echo -n " * Looking for remote database server"

--- a/utils/entrypoint.sh
+++ b/utils/entrypoint.sh
@@ -180,7 +180,7 @@ zm_db_exists() {
 # or mysqld --initialize seems to be wether mysql_install_db is a shell
 # script or a binary executable
 use_mysql_install_db () {
-    MYSQL_INSTALL_DB=$(type mysql_install_db)
+    MYSQL_INSTALL_DB=$(type -p mysql_install_db)
     local result="$?"
 
     if [ "$result" -eq "0" ] && [ -n "$MYSQL_INSTALL_DB"  ]; then

